Show realtime visual graph of profiling

mrbuds opened the pull request:
# Description
Show real time visual graph of profiling
Replace Profiling buttons & bindings

# Ideas for improvement
- resizable

# Current look
![img](https://i.imgur.com/P9nYXnE.gif)
![img](https://i.imgur.com/Ws9tIhc.png)

emptyrivers commented on WeakAuras/Profiling.lua:
```diff
@@ -360,3 +366,160 @@ function WeakAuras.PrintProfile()
   popup:Show()
 end
+
+RealTimeProfilingWindow = CreateFrame("Frame", nil, UIParent)
```
This maybe should be added to WeakAuras.frames.

emptyrivers commented on WeakAuras/Profiling.lua:
```diff
@@ -360,3 +366,160 @@ function WeakAuras.PrintProfile()
+RealTimeProfilingWindow:SetMovable(true)
+
+local texture = "Interface\\DialogFrame\\UI-DialogBox-Background"
+function RealTimeProfilingWindow:AddBar(num, name, map, total)
```
Have you considered using AceGUI instead of raw frame manip?

mrbuds replied:
AceGUI scares me but i'll try, it may help later for adding new buttons

emptyrivers commented on WeakAuras/Profiling.lua:
```diff
@@ -360,3 +366,160 @@ function WeakAuras.PrintProfile()
+
+function RealTimeProfilingWindow:Stop()
+  self:Hide()
+  self:SetScript("OnUpdate", nil)
```
This is implicitly covered by Hide(), since OnUpdate isn’t dispatched to hidden frames.

mrbuds replied:
I didn't know 👍

emptyrivers commented on WeakAuras/Profiling.lua:
```diff
@@ -360,3 +366,160 @@ function WeakAuras.PrintProfile()
+    return
+  end
+
+  local total = TotalProfileTime(WeakAuras.profileData.auras)
```
profileData is an upvalue if this function, so you can just reference it instead of indexing WeakAuras.

emptyrivers commented on WeakAuras/Profiling.lua:
```diff
@@ -360,3 +366,160 @@ function WeakAuras.PrintProfile()
+  self.statsFrameText:SetText(("Time in WeakAuras: %.2f%%"):format(100 * profileData.systems.wa.elapsed / timespent))
+end
+
+function RealTimeProfilingWindow:Start()
```
Here as well you have a lot of raw frame manip that distracts from the purpose of this method. It would be good imo to encapsulate that, with AceGUI widgets or with simple helper functions.

mrbuds replied:
Refactored both RefreshBars() and Start() methods

emptyrivers commented on WeakAuras/Profiling.lua:
```diff
@@ -360,3 +366,160 @@ function WeakAuras.PrintProfile()
+  bar:SetPoint("TOPLEFT", self, "TOPLEFT", 0, - (num - 1) * self.barHeight - self.titleHeight)
+  bar.fg:ClearAllPoints()
+  bar.fg:SetPoint("TOPLEFT", bar)
+  bar.fg:SetSize(self.width / 100 * pct, self.barHeight)
```
Consider mixing in the Smoothed progress mixin from blizzard and using smoothed progress. 

It’s optional but I think it could look quite nice.

mrbuds commented:
Updated pictures in description with last changes
I have replaced all minimap clicks & keybindings to toggle that window instead of previous behavior with the idea to make profiling much easier to use for average user.

emptyrivers commented on WeakAuras/Bindings.xml:
```diff
@@ -2,13 +2,7 @@
-  <Binding name="WEAKAURASPRINTPROFILING" Category="ADDONS">
-    WeakAuras.PrintProfile()
+  <Binding name="WEAKAURASPROFILINGTOGGLE" Category="ADDONS">
+    WeakAuras.RealTimeProfilingWindow:Toggle()
```
I think we should keep the print binding.

Stanzilla replied:
yea

mrbuds replied:
done

Stanzilla commented:
Can you make the bars a bit darker? The white <> yellow has a pretty low contrast

mrbuds commented:
> 
> 
> Can you make the bars a bit darker? The white <> yellow has a pretty low contrast

https://i.imgur.com/ys2y4lU.png ?
